File: python/visa_lk_perks_api_docs/shard.py
"""Shard helpers for parallel probes / crawls.

Split page ranges, CEB groups, or offset windows across workers.
Generated — regenerate via scripts/scaffold-client-extras.py
"""

from __future__ import annotations

from typing import TypeVar

T = TypeVar("T")

_CEB_GROUPS = list("ABCDEFGHIJKLMNOPQRSTUVWXY")


def shard_slice(items: list[T], shard_index: int, shard_count: int) -> list[T]:
    """Return the slice of ``items`` owned by ``shard_index`` (0-based)."""
    if shard_count < 1:
        raise ValueError("shard_count must be >= 1")
    if shard_index < 0 or shard_index >= shard_count:
        raise ValueError("shard_index out of range")
    return [item for i, item in enumerate(items) if i % shard_count == shard_index]


def shard_range(
    start: int,
    end: int,
    shard_index: int,
    shard_count: int,
) -> tuple[int, int]:
    """Inclusive-exclusive ``[lo, hi)`` page/index range for this shard."""
    if end < start:
        raise ValueError("end must be >= start")
    if shard_count < 1:
        raise ValueError("shard_count must be >= 1")
    total = end - start
    base = total // shard_count
    rem = total % shard_count
    lo = start + shard_index * base + min(shard_index, rem)
    hi = lo + base + (1 if shard_index < rem else 0)
    return lo, hi


def shard_page_numbers(
    start_page: int,
    end_page: int,
    shard_index: int,
    shard_count: int,
) -> list[int]:
    """Page numbers (inclusive end) for this shard."""
    lo, hi = shard_range(start_page, end_page + 1, shard_index, shard_count)
    return list(range(lo, hi))


def shard_offsets(
    total_records: int,
    page_size: int,
    shard_index: int,
    shard_count: int,
) -> list[int]:
    """ArcGIS-style resultOffset values for this shard."""
    if page_size < 1:
        raise ValueError("page_size must be >= 1")
    offsets = list(range(0, max(total_records, 0), page_size))
    return shard_slice(offsets, shard_index, shard_count)


def shard_groups(
    shard_index: int,
    shard_count: int,
    groups: list[str] | None = None,
) -> list[str]:
    """CEB LoadShedGroupId letters (default A–Y) for this shard."""
    return shard_slice(list(groups or _CEB_GROUPS), shard_index, shard_count)


def plan_shards(shard_count: int) -> list[dict[str, int]]:
    """Describe shard workers as ``[{index, count}, ...]``."""
    return [{"index": i, "count": shard_count} for i in range(shard_count)]


__all__ = [
    "plan_shards",
    "shard_groups",
    "shard_offsets",
    "shard_page_numbers",
    "shard_range",
    "shard_slice",
]

File: typescript/src/shard.ts
/** Shard helpers for parallel probes / crawls. */

const CEB_GROUPS = "ABCDEFGHIJKLMNOPQRSTUVWXY".split("");

export function shardSlice<T>(items: T[], shardIndex: number, shardCount: number): T[] {
  if (shardCount < 1) throw new Error("shardCount must be >= 1");
  if (shardIndex < 0 || shardIndex >= shardCount) throw new Error("shardIndex out of range");
  return items.filter((_, i) => i % shardCount === shardIndex);
}

export function shardRange(
  start: number,
  end: number,
  shardIndex: number,
  shardCount: number,
): [number, number] {
  if (end < start) throw new Error("end must be >= start");
  if (shardCount < 1) throw new Error("shardCount must be >= 1");
  const total = end - start;
  const base = Math.floor(total / shardCount);
  const rem = total % shardCount;
  const lo = start + shardIndex * base + Math.min(shardIndex, rem);
  const hi = lo + base + (shardIndex < rem ? 1 : 0);
  return [lo, hi];
}

export function shardPageNumbers(
  startPage: number,
  endPage: number,
  shardIndex: number,
  shardCount: number,
): number[] {
  const [lo, hi] = shardRange(startPage, endPage + 1, shardIndex, shardCount);
  return Array.from({ length: hi - lo }, (_, i) => lo + i);
}

export function shardOffsets(
  totalRecords: number,
  pageSize: number,
  shardIndex: number,
  shardCount: number,
): number[] {
  if (pageSize < 1) throw new Error("pageSize must be >= 1");
  const offsets: number[] = [];
  for (let o = 0; o < Math.max(totalRecords, 0); o += pageSize) offsets.push(o);
  return shardSlice(offsets, shardIndex, shardCount);
}

export function shardGroups(
  shardIndex: number,
  shardCount: number,
  groups: string[] = CEB_GROUPS,
): string[] {
  return shardSlice(groups, shardIndex, shardCount);
}

export function planShards(shardCount: number): Array<{ index: number; count: number }> {
  return Array.from({ length: shardCount }, (_, index) => ({ index, count: shardCount }));
}
